// pixi-drawable-textinput/dist/TextInput/InputOption.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var InputFontStyle_1 = require("./InputFontStyle");
var InputCursorStyle_1 = require("./InputCursorStyle");
var InputOption = (function () {
    function InputOption() {
        this.backgroundColor = { color: 0xffffff, alpha: 0.5 };
        this.backgroundColorFocus = { color: 0xffffff, alpha: 1 };
        this.width = 120;
        this.height = 25;
        this.paddingHeight = 10;
        this.style = new InputFontStyle_1.InputFontStyle();
        this.placeHolder = "";
        this.value = "";
        this.maxLength = -1;
        this.roundedBorder = 0;
        this.borderColor = { color: 0, alpha: 1 };
        this.borderColorFocus = { color: 0, alpha: 1 };
        this.borderWidth = 2;
        this.multiLine = false;
        this.wordWrapType = "break-word";
        this.onlyNumber = false;
        this.min = Number.MIN_VALUE;
        this.max = Number.MAX_VALUE;
        this.cursorStyle = new InputCursorStyle_1.InputCursorStyle();
        this.enableCursor = true;
    }
    return InputOption;
}());
exports.InputOption = InputOption;
